upcoming:[DI-23542] - Resource selection limitation phase 2 in cloudpulse alerting resources section

venkymano-akamai opened the pull request:
## Description 📝
Enhance the Resources section in cloudpulse alerting create and edit user alert flow with limitations on maximum number of elements that needs to be selected.

## Changes  🔄

1. Enhance DisplayAlertResources.tsx to disable the root checkbox and item level checkbox based on maximum selections allowed.
2. Enhance AlertResourceNotice.tsx to disable / enable of Select All and Deselect All based on maximum selections allowed.

## Target release date 🗓️

7-04-2025 in Alpha environment

## Preview 📷

| Before  | After   |
| ------- | ------- |
|![Screenshot 2025-03-27 at 9 52 55 AM](https://github.com/user-attachments/assets/9645ac77-25c4-40e8-9b4f-dec2addfdc94)|![Screenshot 2025-03-27 at 9 53 13 AM](https://github.com/user-attachments/assets/7b48281a-289a-4b8c-99c1-ac3e6b3ff36a)|
|![Screenshot 2025-03-27 at 9 53 15 AM](https://github.com/user-attachments/assets/65a71ce3-509b-4598-ac15-b3235c2d2fa3)|![Screenshot 2025-03-27 at 9 53 22 AM](https://github.com/user-attachments/assets/f7269db3-6a5d-48b5-b69b-b7741f8ba1b3)|

## How to test 🧪

1. Login as mock , since some endpoints are not available.
2. In the serverHandler.ts, in databases/instances, make the mock to return around 110 resources 
`return HttpResponse.json(makeResourcePage(databaseInstanceFactory.buildList(110)));`.
3. Navigate to alerts tab inside monitor section.
4. Click on create alerts and select databases as service type.
5. You should be able to see the checkboxes getting disabled after you select 100 resources
6. You should also be able to see the Select All button in notice disabled.

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>

- [x] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [x] All unit tests are passing
- [x] TypeScript compilation succeeded without errors
- [x] Code passes all linting rules

</details>

jaalah-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsResources/DisplayAlertResources.tsx:
```diff
@@ -149,6 +162,17 @@ export const DisplayAlertResources = React.memo(
+      );
+
+      return selectionsRemaining < uncheckedData.length;
+    };
```
Currently, we have two separate implementations:

1. For the header/root checkbox, we use a `disableRootCheckBox` function that filters unchecked data and compares it with `selectionsRemaining`
2. For individual checkboxes, we use a simple inline check: `const boxDisabled = !checked && selectionsRemaining === 0`

We should consolidate this into a single function that handles both cases, maybe something like: 

```
// Usage
isCheckboxDisabled(isChecked: boolean, uncheckedCount?: number)
```

Essentially, let's narrow the parameters needed (rather than passing the entire dataset)

`const uncheckedData = paginatedData.filter(({ checked = false }) => !checked);` can exist outside of the function.

Let me know if you need more clarification

coliu-akamai reviewed (approved):
✅ confirmed max selections tooltip and functionality

thanks @venkymano-akamai 

Unrelated to this PR - can we align the 'Show Selected Only' checkbox with the other elements?

![image](https://github.com/user-attachments/assets/1150b88d-790c-4c08-af05-00476e48bb58)

jaalah-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsResources/DisplayAlertResources.tsx:
```diff
@@ -168,24 +192,53 @@ export const DisplayAlertResources = React.memo(
+                          <Checkbox
+                            indeterminate={
+                              isSomeSelected(paginatedData) &&
+                              !isAllPageSelected(paginatedData)
```
Nit: Optional names for clarity:
- `isSomeSelected` > `hasSelectedItems`
- `isAllPageSelected` > `areAllItemsSelected`

venkymano-akamai replied:
will take this in upcoming work

venkymano-akamai commented:
@coliu-akamai , I have fixed the alignment issues
@jaalah-akamai , I have also fixed by moving it to one function for all the logic like suggested, please go through and check if it is good.

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsResources/DisplayAlertResources.tsx:
```diff
@@ -162,120 +191,182 @@ export const DisplayAlertResources = React.memo(
-                          )
-                        }
+        }) => {
+          const isRootCheckBoxDisabled = isCheckboxDisabled(
+            false,
+            paginatedData.filter(({ checked }) => !checked).length
+          );
```
only once calculate root check box disabled with paginated data and return the table based on it

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsResources/DisplayAlertResources.tsx:
```diff
@@ -162,120 +191,182 @@ export const DisplayAlertResources = React.memo(
+                  {!isDataLoadingError &&
+                    paginatedData.map((resource, index) => {
+                      const { checked, id } = resource;
+                      const boxDisabled = isCheckboxDisabled(checked);
```
calculate current checkbox disabled using same function

jaalah-akamai replied:
nit: `boxDisabled` could be `isItemCheckboxDisabled`

jaalah-akamai reviewed (approved):
Thanks, this looks good to me 👍

venkymano-akamai commented:
Merging since UT's and all jobs passed, the cypress failure a flaky one and it is not related to this
